`type Tree =string | map[string]string` confuse v

### Describe the bug

a c  error is produced with the following program

```
type Tree = string| map[string]string
fn append(mut tree Tree) {
        print(tree["key"])
}

fn main() {
}
```

### Reproduction Steps

```
xcombelle@iMac xco % cat bug.v
type Tree = string| map[string]string 
fn append(mut tree Tree) {
	print(tree["key"])
}

fn main() {
}
```


```

xcombelle@iMac xco % v up       
Updating V...
V is already updated.
Current V version: V 0.4.2 97f7c3f, timestamp: 2023-11-08 20:43:48 +0200
xcombelle@iMac xco % v -cg bug.v
/tmp/v_501/bug.4782279060700994107.tmp.c:12663:27: error: array subscript is not an integer
        print(main__Tree_str(tree[_SLIT("key")]));
                                 ^~~~~~~~~~~~~
1 error generated.
builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

xcombelle@iMac xco % 

```

### Expected Behavior

a v compiler error or no error

### Current Behavior

a c error is produced

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

Current V version: V 0.4.2 97f7c3f

### Environment details (OS name and version, etc.)

xcombelle@iMac xco % v doctor
V full version: V 0.4.2 93d5c2d.97f7c3f
OS: macos, macOS, 14.1, 23B74
Processor: 8 cpus, 64bit, little endian, Apple M1

getwd: /Users/xcombelle/xco
vexe: /Users/xcombelle/v/v
vexe mtime: 2023-11-08 19:35:41

vroot: OK, value: /Users/xcombelle/v
VMODULES: OK, value: /Users/xcombelle/.vmodules
VTMP: OK, value: /tmp/v_501

Git version: git version 2.39.3 (Apple Git-145)
Git vroot status: weekly.2023.44-85-g97f7c3f6
.git/config present: true

CC version: Apple clang version 15.0.0 (clang-1500.0.40.1)
thirdparty/tcc status: thirdparty-macos-arm64 a668e5a0
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -4319,7 +4319,8 @@ fn (mut c Checker) index_expr(mut node ast.IndexExpr) ast.Type {
 	}
 	is_aggregate_arr := typ_sym.kind == .aggregate
 		&& (typ_sym.info as ast.Aggregate).types.filter(c.table.type_kind(it) !in [.array, .array_fixed, .string, .map]).len == 0
-	if typ_sym.kind !in [.array, .array_fixed, .string, .map] && !typ.is_ptr()
+	if typ_sym.kind !in [.array, .array_fixed, .string, .map]
+		&& (!typ.is_ptr() || typ_sym.kind in [.sum_type, .interface_])
 		&& typ !in [ast.byteptr_type, ast.charptr_type] && !typ.has_flag(.variadic)
 		&& !is_aggregate_arr {
 		c.error('type `${typ_sym.name}` does not support indexing', node.pos)

--- a/vlib/v/checker/tests/index_sumtype_or_interface_params_err.out
+++ b/vlib/v/checker/tests/index_sumtype_or_interface_params_err.out
@@ -1,0 +1,13 @@
+vlib/v/checker/tests/index_sumtype_or_interface_params_err.vv:6:9: error: type `Foo` does not support indexing
+    4 | 
+    5 | fn sum_type_or_interface_as_parameters(mut foo Foo, mut bar Bar) {
+    6 |     _ = foo[0]
+      |            ~~~
+    7 |     _ = bar[0]
+    8 | }
+vlib/v/checker/tests/index_sumtype_or_interface_params_err.vv:7:9: error: type `Bar` does not support indexing
+    5 | fn sum_type_or_interface_as_parameters(mut foo Foo, mut bar Bar) {
+    6 |     _ = foo[0]
+    7 |     _ = bar[0]
+      |            ~~~
+    8 | }

--- a/vlib/v/checker/tests/index_sumtype_or_interface_params_err.vv
+++ b/vlib/v/checker/tests/index_sumtype_or_interface_params_err.vv
@@ -1,0 +1,8 @@
+type Foo = int | map[int]int
+
+interface Bar {}
+
+fn sum_type_or_interface_as_parameters(mut foo Foo, mut bar Bar) {
+	_ = foo[0]
+	_ = bar[0]
+}

